msteams-platform/concepts/bots/bots-files.md

```diff
@@ -0,0 +1,215 @@
+| `declineContext` | Additional context that will be silently transmitted to your bot when the user declines the file. |
+
+### Invoke activity when the user accepts the file
+An invoke activity is sent to your bot if and when the user accepts the file. It contains the OneDrive for Business placeholder URL that your bot can then issue a PUT into to transfer the file contents. The following example shows an abridged version of the invoke activity that your bot will receive:
+
+```json
+{
+  ...
+
+  "name": "fileConsent/invoke",
+  "value": {
+    "type": "fileUpload",
+    "action": "accept",
+    "context": {
+    },
+    "uploadInfo": {
+      "contentUrl": "https://contoso.sharepoint.com/personal/johnadams_contoso_com/Documents/Applications/file_example.txt",
+      "name": "file_example.txt",
+      "uploadUrl": "https://upload.link",
+      "uniqueId": "1150D938-8870-4044-9F2C-5BBDEBA70C8C",
+      "fileType": "txt",
+    }
+  }
+}
+```
+
+Similarly, if the user declines the file, your bot will receive the following event:
+
+```json
+{
+  "value": {
+    "type": "fileUpload",
+    "action": "decline",
+    "context": {
+    }
+  }
+}
+```
+
+## Notifying the user about an uploaded file
+After uploading a file to the user's OneDrive, whether you use the mechanism described above or OneDrive user delegated APIs, you should send a confirmation message to the user. This message should contain a `FileCard` attachment that the user can click on, either to preview it, open it in OneDrive, or download locally.
+
+```json
+{
+  "attachments": [{
+    "contentType": "application/vnd.microsoft.teams.card.file.info",
+    "contentUrl": "https://contoso.sharepoint.com/personal/johnadams_contoso_com/Documents/Applications/file_example.txt",
+    "name": "file_example.txt",
+    "content": {
+      "uniqueId": "1150D938-8870-4044-9F2C-5BBDEBA70C8C",
+      "fileType": "txt",
+    }
+  }]
+}
+```
+
+The following table describes the content properties of the attachment:
+
+| Property | Purpose |
+| --- | --- |
+| `uniqueId` | OneDrive/SharePoint drive item ID. |
+| `fileType` | File type, such as pdf or docx. |
+
+## Basic example in C#
+The following sample shows how you can handle file uploads and send file consent requests in your bot's dialog.
+
+```csharp
+
+// This sample dialog shows two simple flows:
+// 1) A silly example of receiving a file from the user, processing the key elements,
+//    and then constructing the attachment and sending it back.
+// 2) Creating a new file consent card requesting user permission to upload a file.
+private async Task MessageReceivedAsync(IDialogContext context, IAwaitable<object> result)
+{
+    var replyMessage = context.MakeMessage();
+    Attachment returnCard;
+
+    var message = await result as Activity;
+
+    // Check to see if the user is sending the bot a file.
+    if (message.Attachments != null && message.Attachments.Any())
+    {
+        var attachment = message.Attachments.First();
```

Comment:
Do we guarantee that the file attachment is always first?

Reply:
Should be in the vast majority of cases. I don't think there will be an attachment otherwise.

Reply:
Hmm, I just checked, and it's easy to get it to not be the first (or only) attachment. I don't think we guarantee the order.
Can you perhaps change this to:
```
var fileAttachment = message.Attachments?.Where(a => a.ContentType == FileDownloadInfo.ContentType)?.FirstOrDefault();
if (fileAttachment != null) {
...
}
```